amethyst_rendy/src/bundle.rs

```diff
@@ -0,0 +1,950 @@
+                let mut framebuffer_layers = u16::max_value();
+
+                for color in colors {
+                    match color {
+                        OutputColor::Surface(surface, _) => {
+                            if let Some(extent) = surface.extent(physical_device) {
+                                framebuffer_width = min(framebuffer_width, extent.width);
+                                framebuffer_height = min(framebuffer_height, extent.height);
+                                framebuffer_layers = min(framebuffer_layers, 1);
+                            } else {
+                                // Window was just closed, using size of 1 is the least bad option
+                                // to default to. The output won't be used, things won't crash and
+                                // graph is either going to be destroyed or rebuilt next frame.
+                                framebuffer_width = min(framebuffer_width, 1);
+                                framebuffer_height = min(framebuffer_height, 1);
+                            }
+                            framebuffer_layers = min(framebuffer_layers, 1);
+                        }
+                        OutputColor::Image(options) => {
+                            let extent = options.kind.extent();
+                            framebuffer_width = min(framebuffer_width, extent.width);
+                            framebuffer_height = min(framebuffer_height, extent.height);
+                            framebuffer_layers = min(framebuffer_layers, options.kind.num_layers());
+                        }
+                    };
+                }
+                if let Some(options) = depth {
+                    let extent = options.kind.extent();
+                    framebuffer_width = min(framebuffer_width, extent.width);
+                    framebuffer_height = min(framebuffer_height, extent.height);
+                    framebuffer_layers = min(framebuffer_layers, options.kind.num_layers());
+                }
+                TargetMetadata {
+                    width: framebuffer_width,
+                    height: framebuffer_height,
+                    layers: framebuffer_layers,
+                }
+            })
+    }
+
+    fn set_outputs(&mut self, outputs: TargetPlanOutputs<B>) -> Result<(), Error> {
+        if self.outputs.is_some() {
+            return Err(format_err!("Target {:?} already defined.", self.key));
+        }
+        self.outputs.replace(outputs);
+        Ok(())
+    }
+
+    fn add_extension(
+        &mut self,
+        extension: Box<dyn FnOnce(&mut TargetPlanContext<'_, B>) -> Result<(), Error> + 'static>,
+    ) {
+        self.extensions.push(extension);
+    }
+
+    fn evaluate(self, ctx: &mut PlanContext<B>) -> Result<(), Error> {
+        if self.outputs.is_none() {
+            return Err(format_err!(
+                "Trying to evaluate not fully defined pass {:?}. Missing `define_pass` call.",
```

Comment:
Won't you hit this if plugin registers pipeline to write into `Custom("VeryImportantTarget")` which not get used anywhere?

Reply:
I can't see any calls to `define_pass` except one that `RenderToWindow` is doing and in tests.

Reply:
There is only one pass in default plugins right now. There will be another once shadow mapping is implemented. You won't hit an error just by extending nonexisting target. It will never be actually evaluated, as the whole system is lazy. If you depend on a target that is never defined, then it is an error and above error would be hit.